core/src/main/java/org/infinispan/xsite/irac/IracResponseCollector.java

```diff
@@ -1,136 +1,111 @@
-      try {
-         if (throwable != null) {
-            exceptionReceived = true;
-            if (DefaultTakeOfflineManager.isCommunicationError(throwable)) {
-               //in case of communication error, we need to back-off.
-               RESULT_UPDATED.set(this, IracBatchSendResult.BACK_OFF_AND_RETRY);
-            } else if (result == IracBatchSendResult.OK) {
-               //don't overwrite communication errors
-               RESULT_UPDATED.compareAndSet(this, IracBatchSendResult.OK, IracBatchSendResult.RETRY);
-            }
-            if (backup.logExceptions()) {
-               log.warnXsiteBackupFailed(cacheName, backup.getSiteName(), throwable);
-            } else if (trace) {
-               log.tracef(throwable, "[IRAC] Encountered issues while backing up data for cache %s to site %s", cacheName, backup.getSiteName());
-            }
-         } else {
-            if (trace) {
-               log.tracef("[IRAC] Received response from site %s (%d missing): %s", backup.getSiteName(), countDownRunnable.missing(), rspIntSet);
-            }
-            mergeIntSetResult(rspIntSet);
-            // if some keys failed to apply, we need to retry.
-            if (!rspIntSet.isEmpty() && result == IracBatchSendResult.OK) {
-               RESULT_UPDATED.compareAndSet(this, IracBatchSendResult.OK, IracBatchSendResult.RETRY);
-            }
+      if (throwable != null) {
+         if (DefaultTakeOfflineManager.isCommunicationError(throwable)) {
+            //in case of communication error, we need to back-off.
+            backup.enableBackOffMode();
+            result = IracBatchSendResult.BACK_OFF_AND_RETRY;
+         } else if (result == IracBatchSendResult.OK) {
+            //don't overwrite communication errors
+            result = IracBatchSendResult.RETRY;
+         }
+         if (backup.logExceptions()) {
+            log.warnXsiteBackupFailed(cacheName, backup.getSiteName(), throwable);
+         } else if (trace) {
+            log.tracef(throwable, "[IRAC] Encountered issues while backing up data for cache %s to site %s", cacheName, backup.getSiteName());
+         }
+         allKeysFailed();
+      } else {
+         if (trace) {
+            log.tracef("[IRAC] Received response from site %s: %s", backup.getSiteName(), rspIntSet);
+         }
+         mergeIntSetResult(rspIntSet);
+         // if some keys failed to apply, we need to retry.
+         if (rspIntSet != null && !rspIntSet.isEmpty() && result == IracBatchSendResult.OK) {
+            result = IracBatchSendResult.RETRY;
          }
-      } finally {
-         countDownRunnable.decrement();
       }
+
+      run();
    }
 
-   @Override
-   public void run() {
-      if (exceptionReceived) {
-         batch.forEach(IracManagerKeyState::retry);
-         listener.onResponseCompleted(result, Collections.emptyList());
-         completableFuture.complete(null);
-         return;
-      }
+   private void run() {
```

Comment:
wouldn't mind to copy the method code to the above method and remove  `run`, which means nothing at this point :)